For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147148

This PR also fixes BMM, which was silently failing for a while.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
check if config.autotune_fallback_to_aten before using aten as a fallback

[ghstack-poisoned]

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/kernel/bmm.py:
```diff
@@ -195,7 +195,11 @@ def may_require_contiguous(t, meta_t):
+        len(choices) == 0
+        and not use_aten_gemm_kernels()
+        and inductor_config.autotune_fallback_to_aten
+    ):
```
super nit - since we use the same condition multiple times, can we make a utility function for it? e.g. we could add one in mm_common.py

henrylhtsang replied:
I agree, will add

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented on torch/_inductor/kernel/mm.py:
```diff
@@ -513,10 +511,8 @@ def tuned_int_mm(mat1, mat2, *, layout=None):
+    if should_fallback_to_aten(choices):
         choices = [aten__int_mm.bind((mat1, mat2), layout)]
 
     try:
```
@chenyang78 I think this try catch is not needed anymore. I plan on removing them in a separate PR, since they are higher risk imo

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -11,12 +11,28 @@
 log = logging.getLogger(__name__)
 
 
+def should_fallback_to_aten(choices) -> bool:
```
What's the type of the input here? At least Sized right?

Skylion007 commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -1161,7 +1161,16 @@ def _shape_match(
-        return X.get_size()[1] == W.get_size()[0]
+        X_size, W_size = X.get_size(), W.get_size()
+        if len(X_size) != len(W_size):
+            log.info("X and W have different ranks")
```
```suggestion
            log.info("X and W have different ranks: %d, %d", len(X_size), len(W_size))
```
This would be effectively free if you cached the lens anyway?

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -10,13 +10,29 @@
+    fallback_to_aten: bool = (
+        len(choices) == 0
+        and not use_aten_gemm_kernels()
+        and inductor_config.autotune_fallback_to_aten
```
Do we need `autotune_fallback_to_aten` when `max_autotune_gemm_backends` exists ? isnt this duplicative ?

henrylhtsang replied:
> Do we need `autotune_fallback_to_aten` when `max_autotune_gemm_backends` exists ? isnt this duplicative ?

I think the intent is to be super safe. Even if users specify only "CUTLASS" and CUTLASS fails, it will still keep things running. I think this safe logic predates the autotune_fallback_to_aten config. 

But yeah it has been pretty painful for me when working on cutlass since bunch of stuff are failing silently.

eellison replied:
If the user wants to fallback, they should do `max_autotune_gemm_backends="CUTLASS, ATEN"`. I dont think we should have two ways of doing the same exact thing.

henrylhtsang replied:
@eellison I agree with the idea. I can commit to removing autotune_fallback_to_aten all together and removing the silent fallback logic. But it will take a few PRs and a while. Does that sound good?

eellison commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -1161,7 +1161,16 @@ def _shape_match(
+        if len(X_size) != len(W_size):
+            log.info("X and W have different ranks: %d, %d", len(X_size), len(W_size))
+            return False
+        if len(X_size) == 2:
+            return X_size[1] == W_size[0]
```
Why do we need this check ? this should already be a precondition of lowering.

henrylhtsang replied:
> Why do we need this check ? this should already be a precondition of lowering.

The _shape_match in CUTLASS2x is a bit different for sparse. So I guess that was the intention.

Let me know what you think. I am fine with removing it here for cutlass 3x.

henrylhtsang replied:
@eellison updated

eellison commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -1161,7 +1161,16 @@ def _shape_match(
+            return X_size[1] == W_size[0]
+        if len(X_size) == 3:
+            # for bmm
+            return X_size[2] == W_size[1]
```
same here

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -126,11 +128,11 @@
-    # The only difference between the two templates is M >= BLOCK_M and N >= BLOCK_N checking.
-    # See more details in https://github.com/pytorch/pytorch/pull/146293
-    else r"""
+        if torch.version.hip is None
```
Unintended format changes ?

henrylhtsang replied:
> Unintended format changes ?

yeah let me remove that..

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison reviewed (commented):
please add test for the newly passing bmm case

henrylhtsang commented:
> please add test for the newly passing bmm case

there is an existing test, `test_max_autotune_cutlass_backend_simple_bmm`

henrylhtsang commented:
@pytorchbot merge

henrylhtsang committed:
Update on "For addmm and bmm, check if config.autotune_fallback_to_aten before using aten as a fallback. Also fix bmm cutlass backend "


This PR also fixes BMM, which was silently failing for a while.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i
